kernel: Fix sleep timing accuracy

Thog opened the pull request:
This PR corrects some mistake while comparing my reversing of kernel 13.x with our own.

``WaitAndCheckScheduledObjects`` timing accuracy was also improved.

gdkchan commented:
Could potentially unblock #1798.

gdkchan reviewed (approved):
lgtm, thanks.

riperiperi commented:
Greatly improves loading speeds in BOTW 1.0.0. Before they were very long and inconsistent, now they are short (3-5x faster) and consistent. This will likely also help ingame streaming, and I noticed a small performance boost ingame. (36-38 -> 40-42)

I took some rough measurements from booting the game to loading a save. "To Load Game" is the time until the save game starts loading (spamming a), "To 100% progress" is the time for the BOTW logo to fill entirely on the loading screen, and "To Ingame" is the time from that to the game beginning to fade in.

I noticed that with a hack to make the game run at 30 instead of 20, "to ingame" shortens to about 5.5 seconds. 

## PR

to load game: 15.42s
to 100% progress: 21.85s
to ingame: 8.74s

to load game: 14.78s
to 100% progress: 22.38s
to ingame: 8.89s

to load game: 14.55s
to 100% progress: 22.18s
to ingame: 8.89s

## Master

to load game: 15.00s
to 100% progress: 1:01.00s
to ingame: 9.58s

to load game: 15.34s
to 100% progress: 1:36.42s
to ingame: 15.27s

to load game: 14.37s
to 100% progress: 59.75s
to ingame: 7.96s

Mou-Ikkai commented:
Hatsune Miku: Project DIVA Megamix seems to no longer crash for the first time on boot, forcing you to press stop emulation and rerun! 
Master: 10/10 times it froze on first run, but ran properly after "Stop Emulation" and relaunching the game.
PR: 0/10 times it froze on first run. This has fixed the first-time startup crash!

jojorne commented:
Tested. Fixed boot on all my games.

riperiperi commented on Ryujinx.HLE/HOS/Kernel/Common/KTimeManager.cs:
```diff
@@ -74,14 +76,19 @@ private void WaitAndCheckScheduledObjects()
+                            if (ms > 0)
+                            {
+                                _waitEvent.WaitOne(ms);
+                            }
```
Maybe should SpinWait a little when ms is 0, so that the critical section lock isn't hammered so hard? Not sure how much is responsive enough.

Thog replied:
Not sure...

vickyorlo commented:
Incidentally, this fixes #2744, and all the other issues in Hyrule Warriors like the timer running slow, various mission progression softlocks/lagginess, and enemy spawning speed. However it feels like the game's performance has dropped - I'm getting framerates below 50 in busy scenes (which, considering it's a musou game, is almost always) when I used to see 60fps with drops to maybe 55 most of the time. It's hard to say, since with this fix the amount of enemies on screen has increased.

riperiperi commented on Ryujinx.HLE/HOS/Kernel/Common/KTimeManager.cs:
```diff
@@ -61,27 +65,49 @@ public void UnscheduleFutureInvocation(IKFutureSchedulerObject schedulerObj)
 
         private void WaitAndCheckScheduledObjects()
         {
+            SpinWait spinWait = new SpinWait();
```
Since this is defined here, it might make sense to reset the spin wait when the spin loop exists, so that it won't yield on its first spin when doing another wait later on.

riperiperi commented on Ryujinx.HLE/HOS/Kernel/Common/KTimeManager.cs:
```diff
@@ -8,6 +8,8 @@ namespace Ryujinx.HLE.HOS.Kernel.Common
 {
     class KTimeManager : IDisposable
     {
+        public static readonly long DefaultTimeIncrementNanoSeconds = ConvertGuestTicksToNanoSeconds(2);
```
`Nanoseconds` rather than `NanoSeconds`?

riperiperi commented on Ryujinx.HLE/HOS/Kernel/Common/KTimeManager.cs:
```diff
@@ -41,11 +44,12 @@ public KTimeManager(KernelContext context)
             lock (_context.CriticalSection.Lock)
             {
                 _waitingObjects.Add(new WaitingObject(schedulerObj, timePoint));
+                Interlocked.Exchange(ref _enforceWakeupFromSpinWait, timeout < 1000000 ? 1 : 0);
```
If something is scheduled with >=1ms timeout, then this will actually _clear the spin wait wakeup, even if there might be scheduled invocation <1 still present in `_waitingObjects`. You could just use an if statement to enforce this rather than the ternary statement.

FICTURE7 replied:
Also is it desirable to do that? I think it only makes sense to interrupt the spin wait if the new time point is smaller than the time point on which the spin wait is waiting on, no?

riperiperi replied:
That is true, but that might be hard to atomically check. The cost of this wakeup is just another run of the `WaitAndCheckScheduledObjects` loop, before it goes back into the spin wait til time is up.

riperiperi commented on Ryujinx.HLE/HOS/Kernel/Common/KTimeManager.cs:
```diff
@@ -61,27 +65,49 @@ public void UnscheduleFutureInvocation(IKFutureSchedulerObject schedulerObj)
                         if (next.TimePoint > timePoint)
                         {
-                            _waitEvent.WaitOne((int)(next.TimePoint - timePoint));
+                            int ms = (int)((next.TimePoint - timePoint) / PerformanceCounter.TicksPerMillisecond);
```
I wonder if 1ms waits will get to this point with 0.99ms remaining, round down to 0 and have to do the spin (maybe similar concerns for longer waits, where it could spin the slightly less than 1ms remainder). I guess it's fine for now, just something to think about. (could a small bias to this calculation help?)

FICTURE7 replied:
Perhaps keep the unit in ticks and convert to ms only in the `WaitOne` call?

d3xMachina commented:
Pokemon Brillant Diamond does not start with this PR. It stay blocked at a black screen even after waiting 5 minutes.
Also if I try to do "Stop emulation" and close Ryujinx while there is the black screen, it does not close Ryujinx unless I close the command line window.

On the latest Ryujinx version it start after like 1 min and the following messages are displayed before loading : 
> 00:00:21.405 |E| HLE.OsThread.9 Gpu WaitOnSyncpoint: Wait on syncpoint 1 for threshold 2 took more than 1000ms, resuming execution...
> 00:01:01.362 |E| HLE.OsThread.9 Gpu WaitOnSyncpoint: Wait on syncpoint 1 for threshold 6 took more than 1000ms, resuming execution...

Here are my logs (not much to see) : 
[Ryujinx_1.0.0+10c19b1.log](https://github.com/Ryujinx/Ryujinx/files/7572775/Ryujinx_1.0.0%2B10c19b1_2021-11-19_21-14-05.-.Copie.log)

The logs after 5 mins are after clicking stop emulation.

d3xMachina commented:
I made a fix for the overflow on the timer.
Also I changed the wait so that the last millisecond we wait is done by spinning even when the wait is higher than 1ms for more precision.
No issue so far.

Here is the patch (rename the extension to .patch) :
[timer_patch.txt](https://github.com/Ryujinx/Ryujinx/files/7573759/timer_patch.txt)

Thog commented:
> I made a fix for the overflow on the timer. Also I changed the wait so that the last millisecond we wait is done by spinning even when the wait is higher than 1ms for more precision. ScheduleFutureInvocation now only interrupt the wait if necessary. No issue so far.
> 
> Here is the patch (rename the extension to .patch) : [timer_patch.txt](https://github.com/Ryujinx/Ryujinx/files/7574436/timer_patch.txt)
> 
> Edit : I added a _waitEvent.Reset() in WaitAndScheduledObject in lock to avoid the case where _waitEvent.Set() is triggered again after _waitEvent.WaitOne() has been interrupted which would cause the next wait to be instantly canceled if higher than 1 ms.
> 
> I didn't completly explain why my patch fix the issue. The bug there was with the code is that _enforceWakeupFromSpinWait was only triggered if another timeout was of less than 1 ms. Instead it should have been canceled when a new scheduled invocation started sooner. With the overflow it would go spinning and only _enforceWakeupFromSpinWait was able to make it go out since the scheduled time was far in the future.

Fixed the issue on https://github.com/Ryujinx/Ryujinx/pull/2828/commits/1f175bf452075b46e6ef7068585afaae7db54228.
I fixed it by removing the unnecessary cast and by capping the upper timeout limit to 0x7FFFFFFF like it used to be. (see ConvertNanosecondsToMilliseconds behaviour)

lostromb commented on Ryujinx.HLE/HOS/Kernel/Common/KTimeManager.cs:
```diff
@@ -119,6 +151,22 @@ public static long ConvertMillisecondsToNanoseconds(long time)
+
+        public static long ConvertGuestTicksToNanoseconds(long ticks)
+        {
+            return (long)Math.Ceiling(ticks * (1000000000.0 / 19200000.0));
```
Where does this 19200000 come from? Is it based on `Stopwatch.Frequency`?

Thog replied:
This is the value of ``CNTFRQ_EL0`` on the Switch
